pkg/discovery/worker/compliance/unschedulable_node_anti_affinity_group_dto_builder.go

```diff
@@ -0,0 +1,100 @@
+// of discovery to buy the 'schedulable' commodity. Without the schedulable commodity the pod continues to stay
+// on the current node unless there are resource constraints or unless moved by the kubernetes controller.
+// In addition, we need pods which do not buy the schedulable commodity to not move to other unschedulable nodes.
+// This is achieved by creating an anti-affinity policy for these pods against other unschedulable nodes.
+type UnschedulableNodesAntiAffinityGroupDTOBuilder struct {
+	cluster  *repository.ClusterSummary
+	targetId string
+	// Manager for tracking schedulable nodes
+	nodesManager *NodeSchedulabilityManager
+}
+
+func NewUnschedulableNodesAntiAffinityGroupDTOBuilder(cluster *repository.ClusterSummary,
+	targetId string, nodesManager *NodeSchedulabilityManager) *UnschedulableNodesAntiAffinityGroupDTOBuilder {
+	return &UnschedulableNodesAntiAffinityGroupDTOBuilder{
+		cluster:      cluster,
+		targetId:     targetId,
+		nodesManager: nodesManager,
+	}
+}
+
+// Creates DTOs for anti-affinity policy between the pods running on a unschedulable nodes to other unschedulable nodes.
+// This will prevent the pods running on a unschedulable nodes to be moved to other other unschedulable nodes.
+func (builder *UnschedulableNodesAntiAffinityGroupDTOBuilder) Build() []*proto.GroupDTO {
+	var policyDTOs []*proto.GroupDTO
+
+	if len(builder.nodesManager.unSchedulableNodes) <= 1 {
+		glog.V(3).Info("Zero or one unschedulable node in the cluster, anti-affinity policies will not be created")
+		return policyDTOs
+	}
+
+	// Map of unschedulable node name to list of UIDs of pods running on these unschedulable nodes
+	nodeToPodsMap := builder.nodesManager.buildNodeNameToPodUIDsMap()
+
+	// List of UIDs of all unschedulable nodes
+	var nodeUIDs []string // UIDs of all unschedulable nodes
+	var nodeUIDtoNameMap = make(map[string]string)
+	for _, nodeName := range builder.nodesManager.unSchedulableNodes {
+		nodeUID := builder.cluster.NodeNameUIDMap[nodeName]
+		nodeUIDs = append(nodeUIDs, nodeUID)
+		nodeUIDtoNameMap[nodeUID] = nodeName
+	}
+
+	// Iterate over all the unschedulable nodes
+	for idx, nodeUID := range nodeUIDs {
+		// policy is being created for the pods on this unschedulable node
+		nodeName := nodeUIDtoNameMap[nodeUID]
+
+		// pod members on the current unschedulable node for the policy
+		podMembers := nodeToPodsMap[nodeName]
+
+		if len(podMembers) == 0 {
+			glog.V(3).Infof("No pods on unschedulable node %s, skipping anti-affinity policy", nodeName)
+			continue
+		}
+
+		// node members comprising of other unschedulable nodes for the policy
+		var otherNodeUIDs []string
+		otherNodeUIDs = append(otherNodeUIDs, nodeUIDs[:idx]...)
+		otherNodeUIDs = append(otherNodeUIDs, nodeUIDs[idx+1:]...)
+
+		nodeMembers := otherNodeUIDs
+
+		// policy id and display name
+		groupID := fmt.Sprintf("UnschedulableNodesAntiAffinity-%s-%s", nodeName, builder.targetId)
```

Comment:
We may want to check if nodeMembers (podMembers as well) is empty and skip the rest of the code if it is.

Reply:
Beginning of the method checks if there is only one unshcedulable node so policy creation can be skipped. Check for podMembers also exists.